### DISCUSSION TITLE
TechArticle、GitHubRepo の集計関数の改善

### DISCUSSION THREAD

**Discussion by nunukim** [lapras-inc/lapras-user-community#24] [スコアについてのご意見・議論]
技術力スコア v2.0 における、記事およびリポジトリの評価の集計の改善点についての議論です。

# 現状

技術力スコア v2.0 において、TechArticleおよびGitHubRepo は、以下の計算式が用いられている。

$$
S = \prod_{n=1}^{\min(3, N)} \log \left(1+ x_n  \right) \\;\\;\\;\\; \cdots (1)
$$

ただし、 $x_n \\, (n=1,\cdots,N)$ はそれぞれ以下の値である。

* GitHubRepo: リポジトリのスター数とコミット数から計算した独自のスコア。 ([ソースコード](https://github.com/lapras-inc/lapras-score-v2/blob/main/lib/lapras_score_v2/calculate_raw_e_score_v2_detail.py#L152))
* TechArticle: 記事のスター数(Zennの場合)またはStock数(Qiitaの場合)。([ソースコード](https://github.com/lapras-inc/lapras-score-v2/blob/be9fc81c2f505bd47fb0746c25ae164371b6f35c/lib/lapras_score_v2/calculate_raw_e_score_v2_detail.py#L233))

これはLAPRASスコアの最初期から残っている実装であり、実装時の思想は以下のようなものである。

* アウトプットの数よりも質を重視するため、アウトプット全件ではなく、上位3件のみを計算対象にしている。
* たまたま1つだけバズったという効果を除外するため、最上位だけでなく、上位3つのアウトプットを計算対象にしている。
* 上位3つの中でも、突出した1件ではなくコンスタントに質の高いアウトプットを出していることを重視する。
  * 突出した値に引っ張られないよう、対数を用いて丸めている。
  * 3件のアウトプットの積（和に比べてより最低値に敏感）を用いている。

# 課題感

上述の現状の実装には、以下のような問題点がある

* 「上位3件」に根拠がない
  * 上位3件だけしか評価しないため、何件もコンスタントに質の高いアウトプットをしている人が評価されづらい
  * 新たにアウトプットした際、自己ベストの上位3位以内に入らないと、スコアに変化を与えない。
* 新たにアウトプットすることでスコアが減ってしまう現象が起こりうる
  * アウトプット数が3件以下の場合、新たなアウトプット($x_n = 1$) が追加されると $\log(1+x_n) < 1$ であるため、全体のスコアが減少してしまう

このような問題を解消するため、現在の計算式に変わる新たな式を定義したい。

# 課題の整理と論点

この課題を一般化すると、 $\\{ x_n \\}$ に対する集合関数 $f(\\{x_n\\})$ を以下のような性質を持つように設計する問題となる。

* アウトプットなしの場合は0点であること $f(\emptyset) = 0$
* 集合関数としての単調増加性: アウトプットをすることでスコアが減ることはないこと $f(X) \leq f(X \cup \\{ x \\})$
* 劣モジュラ性: スコアはアウトプット数に比例せず、新たなアウトプット追加による効果はアウトプット数に応じて逓減すること
$X \subset Y \Rightarrow f(X \cup \\{ x \\} ) - f(X) \leq f(Y \cup \\{ x \\} ) - f(Y)$
* 要素ごとの単調増加性: アウトプット数が同じなら、個々のアウトプットの点数が高い方がスコアが高いこと
$y \leq z \Rightarrow f(x_1, \cdots, y, \cdots, x_N) \leq   f(x_1, \cdots, z, \cdots, x_N)$

これらの性質を満たしながら、アウトプットの評価として妥当な関数 $f$ を設計したい。

設計方針として、以下の2パターンが考えられる。

* 順位ベース: 順位が低いアウトプットの重みを小さくすることで、数よりも質を重視した指標とする。、
  * $x_n$ が降順に並んでいるとして、以下の式で表されるタイプ。
    $$S = \sum_{n=1}^{N} w_n g(x_n)$$
    * ただし、 関数 $g(x) \geq 0$ は単調増加で $g(0) = 0$ 、 $w_n \geq 0$ は単調減少。
  * $w_1 = w_2 = w_3 = 1,  w_n = 0 (n \geq 4)$ として、 $g(x) = \log(\log(1+x))$ とすれば、現在の式(1) の全体に対数をとったものに一致する。
    * ただし、 $g(1) < 0$ のため、単調増加性が壊れてしまっている。(これによりアウトプットすることでスコアが減少する現象が起きている)
  * 現在のロジックを踏襲した単純な解決策として、 $g(x) = \log(1+\log(1+x))$ を用いるなどの解決策はある
    * ただし、上位3件しか見ない課題は残る。
* 値ベース: 値が小さいアウトプットの重みを小さくすることで、数よりも質を重視した指標とする。
  * 順位を気にせずに以下の式で表されるタイプ。(順位ベースにおいて常に $w_n=1$ とした場合に相当)
    $$S = \sum_{n=1}^{N} g(x_n) $$
    * g(x) は、xが大きいほどより寄与が大きくなるようにする(短調増加で下に凸である)。
  * 例えば $g(x) = e^{tx}$ を取ることで、$x_n$ の [LogSumExp](https://en.wikipedia.org/wiki/LogSumExp) を取るのと同等となる。
    * LogSumEXPと一致させるためには全体の対数を取る必要があるが、全体に短調増加関数をかけることは順位に影響を与えない。
    * LogSumExpは近似的に最大値を取ることに対応する。

**Comment by akinomyoga** [comment]
GitHubRepo スコアにおいて

- リポジトリ「上位3件」の決定がどのように行われているかの情報は公開されているでしょうか。特に、評価対象のリポジトリの列挙方法 (別の organization やユーザの下にあるリポジトリは含まれるか & 含む基準は?) と、列挙されたリポジトリを順位付ける基準はどのようになっているでしょうか。
- また、自分(または各ユーザー)の GitHubRepo スコアの決定に実際に用いられた上位3件のリポジトリリストを確認することは可能でしょうか。

https://github.com/lapras-inc/lapras-score-v2/blob/be9fc81c2f505bd47fb0746c25ae164371b6f35c/lib/lapras_score_v2/calculate_raw_e_score_v2_detail.py#L230

を拝見するにこの時点では $`\prod^{\min(3, N)}`$ の上限は設定されていないように見えます。ということは、外部から `github_popular_repos` が入力された時点で既に上位3リポジトリが選別済みということになると思うのですが、そうすると「上位3件」の順位付けは各リポジトリの評価値 $`x_n`$ によるものではなくて、別の基準によるものということでしょうか。

**Reply by rockymanobi** [reply]
@akinomyoga ご質問ありがとうございます！
正確な情報確認してみます！ 水曜日くらいには回答できると思いますので、少々お待ち下さい👍

**Reply by akinomyoga** [reply]
ご対応ありがとうございます! お待ちしております!

**Reply by nunukim** [reply]
数式・コードまで追って質問いただきありがとうございます！

ご指摘の「上位3件」を絞り込む処理につきましては、今回コード公開する際の移植対象から漏れており、LAPRASの本体側のコードに処理がある状態でした。
[こちら](https://github.com/lapras-inc/lapras-score-v2/blob/be9fc81c2f505bd47fb0746c25ae164371b6f35c/lib/lapras_score_v2/calculate_raw_e_score_v2_detail.py#L99)の `github_popular_repos` を `calculate_raw_e_score_v2_detail()` の引数として渡す時点で上位3件に絞り込まれている前提で動くコードとなっておりました。
LAPRAS本体側でリポジトリの「上位3件」を取得する際のスコアリングには、ご提示いただいた `_get_repo_stats_score` と同じものが使われています。
つまり、LAPRAS本体側と今回公開したコードの両方にこの関数が存在してしまっている状態です。
コードの二重管理状態は好ましくないため、「上位3件」を絞り込む処理もこちらのリポジトリに含める形にしたいと思います。
ご指摘いただきありがとうございました！


その他の質問についてもお答えいたします。

* 評価対象のリポジトリの列挙方法 (別の organization やユーザの下にあるリポジトリは含まれるか & 含む基準は?) 
  * 自分が所有するリポジトリ or 自分がコミットしているリポジトリが対象になります。
  * ただし、自分が所有しないリポジトリをforkしたリポジトリについては、コミット数が3件以下のものは除外しています。
* 列挙されたリポジトリを順位付ける基準はどのようになっているでしょうか。
  * こちらは上述の通り、`_get_repo_stats_score` と同じものが使われています。
* 自分(または各ユーザー)の GitHubRepo スコアの決定に実際に用いられた上位3件のリポジトリリストを確認することは可能でしょうか。
  * [ポートフォリオページ](https://lapras.com/person) の「GitHub」の項目に表示されているリポジトリの並び順が `_get_repo_stats_score` の値で降順にソートしたものと同じものになっています。
  * ただし、pin したり 非表示にした場合は順番が変わりますのでご注意ください。

**Reply by akinomyoga** [reply]
詳細な解説ありがとうございます!

> * 列挙されたリポジトリを順位付ける基準はどのようになっているでしょうか。
>   * こちらは上述の通り、`_get_repo_stats_score` と同じものが使われています。
> * 自分(または各ユーザー)の GitHubRepo スコアの決定に実際に用いられた上位3件のリポジトリリストを確認することは可能でしょうか。
>   * [ポートフォリオページ](https://lapras.com/person) の「GitHub」の項目に表示されているリポジトリの並び順が `_get_repo_stats_score` の値で降順にソートしたものと同じものになっています。

ありがとうございます。pin を外して自分のリポジトリ一覧の数値を式に代入してみると確かに降順になっていました!

余談: 少し気になったのはリポジトリ評価値 $`x_n`$ の計算において、スター数・コミット数が 300 でキャップされているので、コントリビューターの数が、上位のリポジトリ乃至その $`x_n`$ を決める主要因子になっていることです。(でも、これは $`x_n`$ の計算方法であって、計算した $`x_n`$ の集計方法ではないのでオフトピックですね…)

**Reply by nunukim** [reply]
ありがとうございます！
 $x_n$ の値についてのご意見もありがたいです。

> スター数・コミット数が 300 でキャップされているので、コントリビューターの数が、上位のリポジトリ乃至その $x_n$ を決める主要因子になっている

確かに、全ての数字が大きい有名OSSへのコントリビュートの場合、コントリユーター数の勝負になってしまっている可能性がありますね・・・
ちょっとデータを見てみたいと思います。

**Reply by akinomyoga** [reply]
ありがとうございます!

**Reply by nunukim** [reply]
リポジトリの評価関数に関する議論が多くなってきたので #31 に分離しました🙏🏻
続きはこちらでお願いします！

**Reply by ryo-endo** [reply]
@akinomyoga 過去のリリースノートを整えていっておりまして。
v2.1.0のときに akinomyoga さんのフィードバックを受けて、リポジトリ「上位3件」の決定がどのように行われているかのロジックを取り込むことができましたので、ささやかながら感謝を込めて コントリビュータにアカウント記載させていただきました。
このたびはありがとうございます！

https://github.com/lapras-inc/lapras-score-v2/releases/tag/v2.1.0

**Reply by akinomyoga** [reply]
ご丁寧にありがとうございます!

**Comment by changeworld** [comment]
GitHub Repo スコアについて

以下の

https://github.com/lapras-inc/lapras-user-community/discussions/24#discussioncomment-11899803

> 評価対象のリポジトリの列挙方法 (別の organization やユーザの下にあるリポジトリは含まれるか & 含む基準は?)
> 自分が所有するリポジトリ or 自分がコミットしているリポジトリが対象になります。
> ただし、自分が所有しないリポジトリをforkしたリポジトリについては、コミット数が3件以下のものは除外しています。

「ただし、自分が所有しないリポジトリをforkしたリポジトリについては、コミット数が3件以下のものは除外しています。」としているのはなぜでしょうか。
「自分が所有しないリポジトリをforkしたリポジトリ」については fork 元でカウントされているので、fork したものについて fork 元と差異がない場合はコミット数に関わらず、除外した方がよいように思えます。

私自身、この算出による影響で私の GitHub Repo スコアの評価が不当に高く評価されてしまっているように見えます。
具体的には、私の GitHub Repo スコアの上位リストを見ると

・OSS1 org/OSS1
・WATASHI/OSS1
・OSS2 org/OSS2
・WATASHI/OSS2

となっていて、OSS1 や OSS2 の評価を 2 重にカウントされています。

・fork 元をカウントするなら、fork したリポジトリについては除外
・fork 元をカウントしないなら、fork したリポジトリをカウントする

という具合にした方がよいと思います。

**Reply by nunukim** [reply]
ご質問・ご意見いただきありがとうございます！
こちらにつきまして、現状仕様の確認のお時間を少々いただきます :pray:（今週中にはお返しできるかと思います）

**Reply by nunukim** [reply]
> 「ただし、自分が所有しないリポジトリをforkしたリポジトリについては、コミット数が3件以下のものは除外しています。」としているのはなぜでしょうか。

こちらにつきまして、OSSをforkだけしたような場合を除外するためのロジックとなっています。

> 「自分が所有しないリポジトリをforkしたリポジトリ」については fork 元でカウントされているので、fork したものについて fork 元と差異がない場合はコミット数に関わらず、除外した方がよいように思えます。

こちらにつきまして、現状のロジックでは「ローカルとfork元のうち、点数が高い方を採用する」という動作を想定していました。
([参考コード](https://github.com/lapras-inc/lapras-score-v2/blob/be9fc81c2f505bd47fb0746c25ae164371b6f35c/lib/lapras_score_v2/calculate_raw_e_score_v2_detail.py#L221))

しかしながら、この処理の前段階で「このリポジトリはこのリポジトリのforkである」という判別ロジックがうまく動作しない場合があることが判明しました。（こちらはLAPRAS本体に実装されています）
どういう条件で発生するのか、調査して対応したいと思います。ご報告ありがとうございました！

（余談ですが、リポジトリ名は一意ではなく変更も可能であったり、GitHub内部で使われているIDも採番され直されてしまうことがあったりと、クローリングの際にリポジトリの同一性を担保するのは意外と難しかったりします・・・）

**Reply by nunukim** [reply]
リポジトリの評価関数に関する議論が多くなってきたので #31 に分離しました🙏🏻
続きはこちらでお願いします！

**Reply by ryo-endo** [reply]
@changeworld forkしたリポジトリが多重計上されている件。
本日リリースしました [LAPRASスコア v.2.2.1 ](https://github.com/lapras-inc/lapras-score-v2/releases/tag/v2.2.1)で修正させていただきました！ 🎉

こちらのフィードバックから不具合の発覚・修正へとつながりました。ありがとうございます。
ささやかではありますが、感謝の気持を込めてリリースノートのコントリビューターにアカウントを載せさせていただきました。

今後ともLAPRASをよろしくお願いいたします。

**Comment by nunukim** [comment]
集計関数ではなくリポジトリの評価関数に関する議論が増えてきたので、これらは #31 にDiscussion分離したいと思います。
